leshan-lwm2m-core/src/main/java/org/eclipse/leshan/core/oscore/OscoreSetting.java

```diff
@@ -99,45 +100,21 @@ public String toString() {
     }
 
     @Override
-    public int hashCode() {
-        final int prime = 31;
-        int result = 1;
-        result = prime * result + ((aeadAlgorithm == null) ? 0 : aeadAlgorithm.hashCode());
-        result = prime * result + ((hkdfAlgorithm == null) ? 0 : hkdfAlgorithm.hashCode());
-        result = prime * result + Arrays.hashCode(masterSalt);
-        result = prime * result + Arrays.hashCode(masterSecret);
-        result = prime * result + Arrays.hashCode(recipientId);
-        result = prime * result + Arrays.hashCode(senderId);
-        return result;
+    public final boolean equals(Object o) {
+        if (this == o)
+            return true;
+        if (!(o instanceof OscoreSetting))
+            return false;
+        OscoreSetting that = (OscoreSetting) o;
+        return Objects.deepEquals(senderId, that.senderId) && Objects.deepEquals(recipientId, that.recipientId)
+                && Objects.deepEquals(masterSecret, that.masterSecret)
+                && Objects.equals(aeadAlgorithm, that.aeadAlgorithm)
+                && Objects.equals(hkdfAlgorithm, that.hkdfAlgorithm) && Objects.deepEquals(masterSalt, that.masterSalt);
     }
 
     @Override
-    public boolean equals(Object obj) {
-        if (this == obj)
-            return true;
-        if (obj == null)
-            return false;
-        if (getClass() != obj.getClass())
-            return false;
-        OscoreSetting other = (OscoreSetting) obj;
-        if (aeadAlgorithm == null) {
-            if (other.aeadAlgorithm != null)
-                return false;
-        } else if (!aeadAlgorithm.equals(other.aeadAlgorithm))
-            return false;
-        if (hkdfAlgorithm == null) {
-            if (other.hkdfAlgorithm != null)
-                return false;
-        } else if (!hkdfAlgorithm.equals(other.hkdfAlgorithm))
-            return false;
-        if (!Arrays.equals(masterSalt, other.masterSalt))
-            return false;
-        if (!Arrays.equals(masterSecret, other.masterSecret))
-            return false;
-        if (!Arrays.equals(recipientId, other.recipientId))
-            return false;
-        if (!Arrays.equals(senderId, other.senderId))
-            return false;
-        return true;
+    public final int hashCode() {
+        return Objects.hash(Arrays.hashCode(senderId), Arrays.hashCode(recipientId), Arrays.hashCode(masterSecret),
+                aeadAlgorithm, hkdfAlgorithm, Arrays.hashCode(masterSalt));
```

Comment:
Do you mean that it should be for example Objects.hash(senderId) and the same for all arrays?

IntelliJ suggests with a warning that: "Array passed to 'Objects.hash()' should be wrapped in 'Arrays.hashcode()". After I leave it without Arrays.hashcode() there is an error in EqualsVerifier:
"Array: regular hashCode() used instead of Arrays.hashCode() for field senderId."

Reply:
> Do you mean that it should be for example Objects.hash(senderId) and the same for all arrays?

Nope clearly we can not do that. If you look at Object.hash() code you will see that Arrays.hashcode() is not called and default hashcode() method of Array is not what we want. (because : https://stackoverflow.com/questions/8777257/equals-vs-arrays-equals-in-java/8777266#8777266)

Generally a valid implementation looks like that : 
```java
    public int hashCode() {
        final int prime = 31;
        int result = 1;
        result = prime * result + Arrays.hashCode(masterSalt);
        result = prime * result + Arrays.hashCode(masterSecret);
        result = prime * result + Arrays.hashCode(recipientId);
        result = prime * result + Arrays.hashCode(senderId);
        result = prime * result + Objects.hash(aeadAlgorithm, hkdfAlgorithm);
        return result;
    }
```
but using `Objects.hash(Arrays.hashCode(senderId), ... )`, this will be equivalent to 
```java
    public int hashCode() {
        final int prime = 31;
        int result = 1;
        result = prime * result + Integer.valueOf(Arrays.hashCode(masterSalt)).hashcode();
        result = prime * result + Integer.valueOf(Arrays.hashCode(masterSecret).hashcode();;
        result = prime * result + Integer.valueOf(Arrays.hashCode(recipientId).hashcode();;
        result = prime * result + Integer.valueOf(Arrays.hashCode(senderId).hashcode();;
        result = prime * result + Objects.hash(aeadAlgorithm, hkdfAlgorithm);
        return result;
    } 
```
And I just don't know if this is OK, maybe it is but I need to know what makes you think we can do that. 

> IntelliJ suggests with a warning that: "Array passed to 'Objects.hash()' should be wrapped in 'Arrays.hashcode()"

This is the kind of information I wanted to know. That makes me think this should be OK to use it.
There is also some [stackoverflow answer ](https://stackoverflow.com/a/30385122/5088764)which goes in that direction (but this is not so argumented :grimacing:)

Looking at : 
 - https://youtrack.jetbrains.com/issue/IDEA-162302/Autogenerating-hashCode-with-java-7-Objects.hashCode-generates-incorrect-code
  - https://youtrack.jetbrains.com/issue/IDEA-179887/IDEA-generates-incorrect-hashCode-methods-for-classes-with-array-fields-when-using-the-template-java.util.Objects.equals-and

There is also some comment which goes in that way but without strong explanation. (maybe this is mathematically obvious but not for me :sweat_smile:)

Anyway, either : 
 -  we trust IntelliJ dev + stackoverlow
 - or we go for a more "classic" way like below : 
 
 ```java
    public int hashCode() {
        final int prime = 31;
        int result = Objects.hash(aeadAlgorithm, hkdfAlgorithm);
        result = prime * result + Arrays.hashCode(masterSalt);
        result = prime * result + Arrays.hashCode(masterSecret);
        result = prime * result + Arrays.hashCode(recipientId);
        result = prime * result + Arrays.hashCode(senderId);
        return result;
    }
 ```
 
 What do you think ? :slightly_smiling_face:

Reply:
To begin with I thought that Objects.hash(Arrays.hashCode()) isn't a problem because it doesn't do the hashing "twice".
I imagined it like this:
1. Arrays.hashCode(...) computes a hash based on array elements
2. Objects.hash(...) does not hash this hash code again (integers have their own hashCode() method that simply returns the integer value?). Instead, it combines it with other hash codes to produce a final hash code

It was only my intuition, I looked at:
https://stackoverflow.com/questions/11890805/hashcode-of-an-int

Reply:
> Objects.hash(...) does not hash this hash code again (integers have their own hashCode() method that simply returns the integer value?). Instead, it combines it with other hash codes to produce a final hash code

I missed that ! I read the code and it seems you're right.

So `Integer.valueOf(Arrays.hashCode(senderId)).hashcode())` = `Arrays.hashCode(senderId)`


So we have now very good reason to keep the elegant code : 

 `Objects.hash(Arrays.hashCode(senderId), Arrays.hashCode(recipientId), Arrays.hashCode(masterSecret),
                aeadAlgorithm, hkdfAlgorithm, Arrays.hashCode(masterSalt)); `
                
:+1: 


Note that unless I missed something,  this also answer to the question about : `Objects.hash (super.hashcode())` ?